llpc/lower/llpcSpirvLowerRayTracingIntrinsics.cpp

```diff
@@ -183,4 +211,108 @@ void SpirvLowerRayTracingIntrinsics::createConvertF32toF16(Function *func, unsig
+  // }
+
+  assert(func->getBasicBlockList().size() == 1);
+  (*func->begin()).eraseFromParent();
+
+  BasicBlock *entryBlock = BasicBlock::Create(m_builder->getContext(), "", func);
+  m_builder->SetInsertPoint(entryBlock);
+  auto argIt = func->arg_begin();
+
+  Value *zero = m_builder->getInt32(0);
+
+  Value *resourcePtr = m_builder->CreateLoad(argIt->getType()->getPointerElementType(), argIt);
+  Value *ptr = m_builder->CreateGEP(resourcePtr->getType()->getPointerElementType(), resourcePtr, {zero});
+
+  m_builder->CreateRet(m_builder->CreatePtrToInt(ptr, m_builder->getInt64Ty()));
+}
+
+// =====================================================================================================================
+// Create AmdExtAtomicFMinAtAddrx2, AmdExtAtomicFMinAtAddr, AmdExtAtomicFMaxAtAddrx2, AmdExtAtomicFMaxAtAddr
+//
+// @param func : Function to create
+// @param isMin : Whether is Min operation, otherwise Max operation
+// @param is64Ty : Whether is 64-bit type operation, otherwise 32-bit type operation
+void SpirvLowerRayTracingIntrinsics::createAtomicFMinMaxAtAddr(Function *func, bool isMin, bool is64Ty) {
+  // float(2) AmdExtAtomicFMinAtAddr(x2)(uint64_t gpuVa, uint offset, float(2) value)
+
+  assert(func->getBasicBlockList().size() == 1);
+  (*func->begin()).eraseFromParent();
+
+  BasicBlock *entryBlock = BasicBlock::Create(m_builder->getContext(), "", func);
+  m_builder->SetInsertPoint(entryBlock);
+  auto argIt = func->arg_begin();
+
+  Type *gpuVaTy = m_builder->getInt64Ty();
+  Type *valueTy = is64Ty ? cast<Type>(FixedVectorType::get(m_builder->getFloatTy(), 2)) : m_builder->getFloatTy();
+  // TODO: Remove this when LLPC will switch fully to opaque pointers.
+  assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(argIt->getType(), gpuVaTy));
+  Value *gpuVa = m_builder->CreateLoad(gpuVaTy, argIt++);
+  Value *offset = m_builder->CreateLoad(m_builder->getInt32Ty(), argIt++);
+  // TODO: Remove this when LLPC will switch fully to opaque pointers.
+  assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(argIt->getType(), valueTy));
+  Value *value = m_builder->CreateLoad(valueTy, argIt);
+
+  if (is64Ty)
+    value = m_builder->CreateBitCast(value, m_builder->getDoubleTy());
+
+  Type *gpuVaAsPtrTy = Type::getInt8PtrTy(m_builder->getContext(), SPIRAS_Uniform);
+  auto gpuVaAsPtr = m_builder->CreateIntToPtr(gpuVa, gpuVaAsPtrTy);
+  // Create GEP to get the byte address with byte offset
+  gpuVaAsPtr = m_builder->CreateGEP(m_builder->getInt8Ty(), gpuVaAsPtr, offset);
+  Type *gpuVaPtrTy = is64Ty ? Type::getDoublePtrTy(m_builder->getContext(), SPIRAS_Uniform)
+                            : Type::getFloatPtrTy(m_builder->getContext(), SPIRAS_Uniform);
+  gpuVaAsPtr = m_builder->CreateBitCast(gpuVaAsPtr, gpuVaPtrTy);
+
+  AtomicRMWInst::BinOp binOp = isMin ? AtomicRMWInst::FMin : AtomicRMWInst::FMax;
+
+  Value *ret =
+      m_builder->CreateAtomicRMW(binOp, gpuVaAsPtr, value, MaybeAlign(), AtomicOrdering::Monotonic, SyncScope::System);
```

Comment:
These functions aim to replace InterlockedMin/InterlockedMax (e.g. [these](https://github.com/GPUOpen-Drivers/gpurt/blob/1f0c4f7e9cea22452e5e20a6cdfc4a84a2bf5bac/src/shaders/BuildBVHTDTR.hlsl#L377)) that operates on uint with float ones, to avoid redundant type conversion. I think `Monotonic` is fine since the original ones also use this.

Reply:
Huh, it's strange that an HLSL `InterlockedMin` would only have monotonic ordering, but okay, if you're sure :)